docs: integrations sidebar updates (#262)
langchain-ai/docs: src/oss/integrations/contributing/standard_tests.mdx

---
pagination_next: contributing/how_to/integrations/publish
pagination_prev: contributing/how_to/integrations/package
title: Add standard tests
---

When creating either a custom class for yourself or to publish in a LangChain integration, it is important to add standard tests to ensure it works as expected. This guide will show you how to add standard tests to each integration type.

## Setup

First, let's install 2 dependencies:

- `langchain-core` will define the interfaces we want to import to define our custom tool.

> - `langchain-tests` will provide the standard tests we want to use, as well as pytest plugins necessary to run them. Recommended to pin to the [latest version](https://pypi.org/project/langchain-tests/).

Reviewer: ```suggestion
- `langchain-tests` will provide the standard tests we want to use, as well as `pytest` plugins necessary to run them. We recommend pinning to the [latest version](https://pypi.org/project/langchain-tests/).
```

<Note>

Because added tests in new versions of `langchain-tests` can break your CI/CD pipelines, we recommend pinning the
version of `langchain-tests` to avoid unexpected changes.

</Note>

<Tabs>
    <Tab title="Poetry">

> If you followed the [previous guide](../package), you should already have these dependencies installed!

Reviewer: ```suggestion
If you followed the [previous guide](../package), you should already have these dependencies installed.
```

```bash
poetry add langchain-core
poetry add --group test langchain-tests==<latest_version>
poetry install --with test
```
    </Tab>
    <Tab title="Pip">
```bash
pip install -U langchain-core langchain-tests

# install current package in editable mode
pip install --editable .
```
    </Tab>
</Tabs>

## Add and configure standard tests

There are 2 namespaces in the `langchain-tests` package:

- Unit tests (`langchain_tests.unit_tests`): designed to be used to test the component in isolation and without access to external services
- Integration tests (`langchain_tests.integration_tests`): designed to be used to test the component with access to external services (in particular, the external service that the component is designed to interact with).

> Both types of tests are implemented as [`pytest` class-based test suites](https://docs.pytest.org/en/7.1.x/getting-started.html#group-multiple-tests-in-a-class).

Reviewer: ```suggestion
Both types of tests are implemented as `pytest` [class-based test suites](https://docs.pytest.org/en/7.1.x/getting-started.html#group-multiple-tests-in-a-class).
```

classes in test files under two test subdirectories:

- `tests/unit_tests` for unit tests
- `tests/integration_tests` for integration tests

### Implementing standard tests

In the following tabs, we show how to implement the standard tests for
each component type:

<Tabs>

    <Tab title="Chat models">

To configure standard tests for a chat model, we subclass `ChatModelUnitTests` and `ChatModelIntegrationTests`. On each subclass, we override the following `@property` methods to specify the chat model to be tested and the chat model's configuration:

| Property | Description |
| --- | --- |
| `chat_model_class` | The class for the chat model to be tested |
| `chat_model_params` | The parameters to pass to the chat
model's constructor |

Additionally, chat model standard tests test a range of behaviors, from the most basic requirements (generating a response to a query) to optional capabilities like multi-modal support and tool-calling. For a test run to be successful:

1. If a feature is intended to be supported by the model, it should pass;
2. If a feature is not intended to be supported by the model, it should be skipped.

> Tests for "optional" capabilities are controlled via a set of properties that can be overridden on the test model subclass.

Reviewer: ```suggestion
Tests for optional capabilities are controlled via a set of properties that can be overridden on the test model subclass.
```

> You can see the **entire list of configurable capabilities** in the API references for

Reviewer: ```suggestion
You can see the entire list of configurable capabilities in the API references for
```

[unit tests](https://python.langchain.com/api_reference/standard_tests/unit_tests/langchain_tests.unit_tests.chat_models.ChatModelUnitTests.html)
and [integration tests](https://python.langchain.com/api_reference/standard_tests/integration_tests/langchain_tests.integration_tests.chat_models.ChatModelIntegrationTests.html).

> For example, to enable integration tests for image inputs, we can implement

Reviewer: ```suggestion
For example, to enable integration tests for image inputs, we can implement:
```

            "model": "bird-brain-001",
            "temperature": 0,
            "parrot_buffer_length": 50,
        }
```

    </Tab>
    <Tab title="Vector stores">

Here's how you would configure the standard tests for a typical vector store (using
`ParrotVectorStore` as a placeholder):

Vector store tests do not have optional capabilities to be configured at this time.

```python title="tests/integration_tests/test_vectorstores.py"
from typing import Generator

import pytest
from langchain_parrot_link.vectorstores import ParrotVectorStore
from langchain_core.vectorstores import VectorStore
from langchain_tests.integration_tests import VectorStoreIntegrationTests

class TestParrotVectorStore(VectorStoreIntegrationTests):
    @pytest.fixture()
    def vectorstore(self) -> Generator[VectorStore, None, None]:  # type: ignore
        """Get an empty vectorstore for unit tests."""
        store = ParrotVectorStore(self.get_embeddings())
        # note: store should be EMPTY at this point
        # if you need to delete data, you may do so here
        try:
            yield store
        finally:
            # cleanup operations, or deleting data
            pass
```

> Configuring the tests consists of implementing pytest fixtures for setting up an

Reviewer: ```suggestion
Configuring the tests consists of implementing `pytest` fixtures for setting up an
```

that will generate short, arbitrary vectors for documents. You can use a different
embeddings object if desired.

In the `finally` block, we call whatever integration-specific logic is needed to
bring the vector store to a clean state. This logic is executed in between each test
(e.g., even if tests fail).

<Note>

Details on what tests are run and troubleshooting tips for each test can be found in the [API reference](https://python.langchain.com/api_reference/standard_tests/integration_tests/langchain_tests.integration_tests.vectorstores.VectorStoreIntegrationTests.html).

</Note>

    </Tab>
    <Tab title="Embeddings">

To configure standard tests for an embeddings model, we subclass `EmbeddingsUnitTests` and `EmbeddingsIntegrationTests`. On each subclass, we override the following `@property` methods to specify the embeddings model to be tested and the embeddings model's configuration:

| Property | Description |
| --- | --- |
| `embeddings_class` | The class for the embeddings model to be tested |
| `embedding_model_params` | The parameters to pass to the embeddings model's constructor |

<Note>

> Details on what tests are run, how each test can be skipped, and troubleshooting tips for each test can be found in the API references. See details:

Reviewer: ```suggestion
Details on what tests are run, how each test can be skipped, and troubleshooting tips for each test can be found in the API references:
```

class TestParrotLinkEmbeddingsIntegration(EmbeddingsIntegrationTests):
    @property
    def embeddings_class(self) -> Type[ParrotLinkEmbeddings]:
        return ParrotLinkEmbeddings

    @property
    def embedding_model_params(self) -> dict:
        return {"model": "nest-embed-001"}
```

    </Tab>
    <Tab title="Tools">

To configure standard tests for a tool, we subclass `ToolsUnitTests` and
`ToolsIntegrationTests`. On each subclass, we override the following `@property` methods
to specify the tool to be tested and the tool's configuration:

| Property | Description |
| --- | --- |
| `tool_constructor` | The constructor for the tool to be tested, or an instantiated tool. |
| `tool_constructor_params` | The parameters to pass to the tool (optional). |
| `tool_invoke_params_example` | An example of the parameters to pass to the tool's `invoke` method. |

If you are testing a tool class and pass a class like `MyTool` to `tool_constructor`, you can pass the parameters to the constructor in `tool_constructor_params`.

If you are testing an instantiated tool, you can pass the instantiated tool to `tool_constructor` and do not
override `tool_constructor_params`.

<Note>

> Details on what tests are run, how each test can be skipped, and troubleshooting tips for each test can be found in the API references. See details:

Reviewer: ```suggestion
Details on what tests are run, how each test can be skipped, and troubleshooting tips for each test can be found in the API references:
```

| `retriever_query_example` | An example of the query to pass to the retriever's `invoke` method |

<Note>

Details on what tests are run and troubleshooting tips for each test can be found in the [API reference](https://python.langchain.com/api_reference/standard_tests/integration_tests/langchain_tests.integration_tests.retrievers.RetrieversIntegrationTests.html).

</Note>

```python title="tests/integration_tests/test_retrievers.py"
from typing import Type

from langchain_parrot_link.retrievers import ParrotRetriever
from langchain_tests.integration_tests import (
    RetrieversIntegrationTests,
)

class TestParrotRetriever(RetrieversIntegrationTests):
    @property
    def retriever_constructor(self) -> Type[ParrotRetriever]:
        """Get an empty vectorstore for unit tests."""
        return ParrotRetriever

    @property
    def retriever_constructor_params(self) -> dict:
        return {"k": 2}

    @property
    def retriever_query_example(self) -> str:
        """
        Returns a str representing the "query" of an example retriever call.
        """
        return "example query"
```

    </Tab>
</Tabs>

---

### Running the tests

> You can run these with the following commands from your project root

Reviewer: ```suggestion
You can run these with the following commands from your project root.
```

<Tabs>
    <Tab title="Poetry" default>

```bash
# run unit tests without network access
poetry run pytest --disable-socket --allow-unix-socket --asyncio-mode=auto tests/unit_tests

# run integration tests
poetry run pytest --asyncio-mode=auto tests/integration_tests
```

    </Tab>
    <Tab title="Pip">

```bash
# run unit tests without network access
pytest --disable-socket --allow-unix-socket --asyncio-mode=auto tests/unit_tests

# run integration tests
pytest --asyncio-mode=auto tests/integration_tests
```

    </Tab>
</Tabs>

## Test suite information and troubleshooting

For a full list of the standard test suites that are available, as well as
information on which tests are included and how to troubleshoot common issues,
see the [Standard Tests API Reference](https://python.langchain.com/api_reference/standard_tests/index.html).

You can see troubleshooting guides under the individual test suites listed in that API Reference. For example,

> [here is the guide for `ChatModelIntegrationTests.test_usage_metadata`](https://python.langchain.com/api_reference/standard_tests/integration_tests/langchain_tests.integration_tests.chat_models.ChatModelIntegrationTests.html#langchain_tests.integration_tests.chat_models.ChatModelIntegrationTests.test_usage_metadata).

Reviewer: ```suggestion
here is the guide for [ChatModelIntegrationTests.test_usage_metadata](https://python.langchain.com/api_reference/standard_tests/integration_tests/langchain_tests.integration_tests.chat_models.ChatModelIntegrationTests.html#langchain_tests.integration_tests.chat_models.ChatModelIntegrationTests.test_usage_metadata).
```
